src/NUnitFramework/framework/Internal/Results/TestSuiteResult.cs

```diff
@@ -112,46 +117,53 @@ public override IList<ITestResult> Children
         /// <param name="result">The result to be added</param>
         public virtual void AddResult(ITestResult result)
         {
-            Children.Add(result);
+            Children.Enqueue(result);
 
             //AssertCount += result.AssertCount;
 
-            // If this result is marked cancelled, don't change it
-            if (ResultState != ResultState.Cancelled)
+            bool stateSet;
+            do
+            {
+                stateSet = true;
+
+                var resultState = ResultState;
+
+                // If this result is marked cancelled, don't change it
+                if (resultState == ResultState.Cancelled)
+                    break;
+
                 switch (result.ResultState.Status)
                 {
                     case TestStatus.Passed:
 
-                        if (ResultState.Status == TestStatus.Inconclusive)
-                            SetResult(ResultState.Success);
+                        if (resultState.Status == TestStatus.Inconclusive)
+                            stateSet = SetResultIf(resultState, ResultState.Success);
 
                         break;
 
                     case TestStatus.Failed:
 
 
-                        if (ResultState.Status != TestStatus.Failed)
-                            SetResult(ResultState.ChildFailure, CHILD_ERRORS_MESSAGE);
+                        if (resultState.Status != TestStatus.Failed)
+                            stateSet = SetResultIf(resultState, ResultState.ChildFailure, CHILD_ERRORS_MESSAGE);
 
                         break;
 
                     case TestStatus.Skipped:
 
                         if (result.ResultState.Label == "Ignored")
-                            if (ResultState.Status == TestStatus.Inconclusive || ResultState.Status == TestStatus.Passed)
-                                SetResult(ResultState.Ignored, CHILD_IGNORE_MESSAGE);
-
-                        break;
+                            if (resultState.Status == TestStatus.Inconclusive || resultState.Status == TestStatus.Passed)
+                                stateSet = SetResultIf(resultState, ResultState.Ignored, CHILD_IGNORE_MESSAGE);
 
-                    default:
                         break;
                 }
+            } while (!stateSet);
 
-            AssertCount += result.AssertCount;
-            _passCount += result.PassCount;
-            _failCount += result.FailCount;
-            _skipCount += result.SkipCount;
-            _inconclusiveCount += result.InconclusiveCount;
+            Interlocked.Add (ref InternalAssertCount, result.AssertCount);
+            Interlocked.Add (ref _passCount, result.PassCount);
+            Interlocked.Add (ref _failCount, result.FailCount);
+            Interlocked.Add (ref _skipCount, result.SkipCount);
+            Interlocked.Add (ref _inconclusiveCount, result.InconclusiveCount);
         }
 
```

Comment:
If we just had a lock around this method, do we even need the concurrent queue? If Children were just IEnumerable, then consumers wouldn't be able to add children and we could enforce everything here greatly simplifying this entire PR.